debian/sandbox-app-launcher.postinst

```diff
@@ -32,6 +31,10 @@ if ! [ -d "${shared_dir}" ]; then
   mkdir -m 1777 "${shared_dir}"
 fi
 
+if ! [ -f "${wrapper_script}-wx" ]; then
+  cp "${wrapper_script}" "${wrapper_script}-wx"
```

Comment:
Can we use a symlink instead?

See for example:

whonix-base-files `debian/whonix-base-files.links`

Better to keep this file updated?

Reply:
AppArmor doesn't seem to apply the policy correctly to symlinks:

```
user@host:/usr/share/sandbox-app-launcher$ sudo ln -s wrapper-script wrapper-script-wx
user@host:/usr/share/sandbox-app-launcher$ sandbox-app-launcher start bash
Failed to check for virtualization: Permission denied
bash: cannot set terminal process group (1): Inappropriate ioctl for device
bash: no job control in this shell
sal-bash@host:~$ ps auxZ
LABEL                           USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
sandbox-app-launcher (enforce)  sal-bash     3  0.0  0.1   6776  3492 ?        S    20:30   0:00 /bin/bash /usr/share/sandbox-app-launcher/wrapper-script-wx
sandbox-app-launcher (enforce)  sal-bash     9  0.0  0.1   8556  2536 ?        Ss   20:30   0:00 /usr/bin/dbus-daemon --syslog --fork --print-pid 5 --print-ad
sandbox-app-launcher (enforce)  sal-bash    10  0.0  0.1   6992  3680 ?        S    20:30   0:00 /bin/bash
sandbox-app-launcher (enforce)  sal-bash    11  0.0  0.1  13304  3484 ?        R    20:30   0:00 ps auxZ
sal-bash@host:~$
```

Whereas when copying the file:

```
user@host:/usr/share/sandbox-app-launcher$ sudo cp wrapper-script wrapper-script-wx
user@host:/usr/share/sandbox-app-launcher$ sandbox-app-launcher start bash
Failed to check for virtualization: Permission denied
bash: cannot set terminal process group (1): Inappropriate ioctl for device
bash: no job control in this shell
sal-bash@host:~$ ps auxZ
LABEL                           USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
sandbox-app-launcher-wx (enforce) sal-bash   3  0.0  0.1   6776  3488 ?        S    20:31   0:00 /bin/bash /usr/share/sandbox-app-launcher/wrapper-script-wx
sandbox-app-launcher-wx (enforce) sal-bash   9  0.0  0.1   8556  2536 ?        Ss   20:31   0:00 /usr/bin/dbus-daemon --syslog --fork --print-pid 5 --print-ad
sandbox-app-launcher-wx (enforce) sal-bash  10  0.0  0.1   6992  3708 ?        S    20:31   0:00 /bin/bash
sandbox-app-launcher-wx (enforce) sal-bash  11  0.0  0.1  13304  3484 ?        R    20:31   0:00 ps auxZ
sal-bash@host:~$
```

(the other errors are fine; they're just because of `/etc/X11/Xsession.d/20software_rendering_in_vms` and bwrap's `--new-session`)

`bash` had the `allow_dynamic_native_code_exec=yes` permission in both cases.

A hardlink (i.e. `ln` without `-s`) seems to work, however. Would that be a good idea instead? Is it even supported?

Reply:
`if ! [ -f "${wrapper_script}-wx" ]; then<br class="Apple-interchange-newline">` means that isn't done if that file already exists. I.e. no updates would be applied. Therefore best to drop that if and do it unconditionally? Could do a diff beforehand and only copy if file was changed but that has very little upside and results in more lines of code.

AppArmor being incompatible with symlinks would be a very strong argument against symlinks indeed here.

Dunno about hardlinks.

I guess symlinks can be made to work. Whonix ships a few anyhow. See for example:

    ls -ls /etc/resolv.conf*

That has existing apparmor profiles. Works. See `apparmor-profile-dist`. Does that work?

Reply:
> Therefore best to drop that if and do it unconditionally?

https://github.com/Whonix/sandbox-app-launcher/pull/62

> I guess symlinks can be made to work. Whonix ships a few anyhow.

It's not that AppArmor doesn't work with symlinks at all, but rather that it applies the policy of the *origin executable* to the *symlink*. For example, if you symlink `file1` -> `symlink1` and specify an AppArmor profile for both files, AppArmor will only enforce the profile for the origin (in this case, `file1`) and ignore the profile for the symlink (in this case, `symlink1`). Any attempts at executing `symlink1` will result in it being enforced under the profile for `file1` and not the profile for `symlink1`.